File: sdk/python/cuttlegate/openfeature.py
"""OpenFeature Provider for Cuttlegate.

Implements the OpenFeature ``AbstractProvider`` class using the
``openfeature-sdk`` package.

Usage::

    from openfeature import api as openfeature
    from cuttlegate import CuttlegateClient
    from cuttlegate.openfeature import CuttlegateProvider

    client = CuttlegateClient(config)
    openfeature.set_provider(CuttlegateProvider(client))

    of_client = openfeature.get_client()
    value = of_client.get_boolean_value("dark-mode", False)
"""

from __future__ import annotations

import json
import typing

from openfeature.evaluation_context import EvaluationContext
from openfeature.flag_evaluation import FlagResolutionDetails, Reason
from openfeature.provider import AbstractProvider, Metadata

from .client import CuttlegateClient
from .types import EvalContext


class CuttlegateProvider(AbstractProvider):
    """OpenFeature provider backed by a sync CuttlegateClient."""

    def __init__(self, client: CuttlegateClient) -> None:
        super().__init__()
        self._client = client

    def get_metadata(self) -> Metadata:
        return Metadata(name="cuttlegate")

    def _eval_context(self, context: typing.Optional[EvaluationContext] = None) -> EvalContext:
        if context is None:
            return EvalContext(user_id="")
        targeting_key = context.targeting_key or ""
        attrs = {k: v for k, v in (context.attributes or {}).items() if isinstance(v, str)}
        return EvalContext(user_id=str(targeting_key), attributes=attrs)

    def resolve_boolean_details(
        self,
        flag_key: str,
        default_value: bool,
        evaluation_context: typing.Optional[EvaluationContext] = None,
    ) -> FlagResolutionDetails[bool]:
        try:
            result = self._client.bool(flag_key, self._eval_context(evaluation_context))
            return FlagResolutionDetails(
                value=result,
                variant="true" if result else "false",
                reason=Reason.TARGETING_MATCH,
            )
        except Exception as exc:
            return FlagResolutionDetails(
                value=default_value,
                reason=Reason.ERROR,
                error_code="GENERAL",
                error_message=str(exc),
            )

    def resolve_string_details(
        self,
        flag_key: str,
        default_value: str,
        evaluation_context: typing.Optional[EvaluationContext] = None,
    ) -> FlagResolutionDetails[str]:
        try:
            result = self._client.string(flag_key, self._eval_context(evaluation_context))
            return FlagResolutionDetails(
                value=result,
                variant=result,
                reason=Reason.TARGETING_MATCH,
            )
        except Exception as exc:
            return FlagResolutionDetails(
                value=default_value,
                reason=Reason.ERROR,
                error_code="GENERAL",
                error_message=str(exc),
            )

    def resolve_integer_details(
        self,
        flag_key: str,
        default_value: int,
        evaluation_context: typing.Optional[EvaluationContext] = None,
    ) -> FlagResolutionDetails[int]:
        try:
            result = self._client.string(flag_key, self._eval_context(evaluation_context))
            return FlagResolutionDetails(
                value=int(result),
                variant=result,
                reason=Reason.TARGETING_MATCH,
            )
        except Exception as exc:
            return FlagResolutionDetails(
                value=default_value,
                reason=Reason.ERROR,
                error_code="GENERAL",
                error_message=str(exc),
            )

    def resolve_float_details(
        self,
        flag_key: str,
        default_value: float,
        evaluation_context: typing.Optional[EvaluationContext] = None,
    ) -> FlagResolutionDetails[float]:
        try:
            result = self._client.string(flag_key, self._eval_context(evaluation_context))
            return FlagResolutionDetails(
                value=float(result),
                variant=result,
                reason=Reason.TARGETING_MATCH,
            )
        except Exception as exc:
            return FlagResolutionDetails(
                value=default_value,
                reason=Reason.ERROR,
                error_code="GENERAL",
                error_message=str(exc),
            )

    def resolve_object_details(
        self,
        flag_key: str,
        default_value: typing.Any,
        evaluation_context: typing.Optional[EvaluationContext] = None,
    ) -> FlagResolutionDetails[typing.Any]:
        try:
            result = self._client.string(flag_key, self._eval_context(evaluation_context))
            return FlagResolutionDetails(
                value=json.loads(result),
                variant=result,
                reason=Reason.TARGETING_MATCH,
            )
        except Exception as exc:
            return FlagResolutionDetails(
                value=default_value,
                reason=Reason.ERROR,
                error_code="GENERAL",
                error_message=str(exc),
            )

File: sdk/js/src/openfeature.ts
/**
 * OpenFeature Provider for Cuttlegate.
 *
 * Implements the `Provider` interface from `@openfeature/server-sdk`.
 *
 * @example
 * ```ts
 * import { OpenFeature } from '@openfeature/server-sdk';
 * import { createClient } from '@cuttlegate/sdk';
 * import { CuttlegateProvider } from '@cuttlegate/sdk/openfeature';
 *
 * const client = createClient({ ... });
 * await OpenFeature.setProviderAndWait(new CuttlegateProvider(client));
 *
 * const ofClient = OpenFeature.getClient();
 * const value = await ofClient.getBooleanValue('dark-mode', false);
 * ```
 */

import type {
  EvaluationContext,
  JsonValue,
  Logger,
  Provider,
  ResolutionDetails,
} from '@openfeature/server-sdk';
import { ErrorCode, ServerProviderStatus } from '@openfeature/server-sdk';
import type { CuttlegateClient } from './client.js';
import type { EvalContext } from './types.js';

/**
 * CuttlegateOpenFeatureProvider implements the OpenFeature `Provider`
 * interface backed by a CuttlegateClient.
 */
export class CuttlegateOpenFeatureProvider implements Provider {
  readonly metadata = { name: 'cuttlegate' } as const;
  status: ServerProviderStatus = ServerProviderStatus.READY;
  readonly hooks = [];

  private client: CuttlegateClient;

  constructor(client: CuttlegateClient) {
    this.client = client;
  }

  private toEvalContext(ctx: EvaluationContext): EvalContext {
    const attributes: Record<string, string> = {};
    for (const [k, v] of Object.entries(ctx)) {
      if (k !== 'targetingKey' && typeof v === 'string') {
        attributes[k] = v;
      }
    }
    return {
      user_id: ctx.targetingKey ?? '',
      attributes,
    };
  }

  async resolveBooleanEvaluation(
    flagKey: string,
    defaultValue: boolean,
    context: EvaluationContext,
    _logger: Logger,
  ): Promise<ResolutionDetails<boolean>> {
    try {
      const result = await this.client.bool(flagKey, this.toEvalContext(context));
      return { value: result, variant: result ? 'true' : 'false', reason: 'TARGETING_MATCH' };
    } catch (err) {
      return {
        value: defaultValue,
        reason: 'ERROR',
        errorCode: ErrorCode.GENERAL,
        errorMessage: err instanceof Error ? err.message : 'Unknown error',
      };
    }
  }

  async resolveStringEvaluation(
    flagKey: string,
    defaultValue: string,
    context: EvaluationContext,
    _logger: Logger,
  ): Promise<ResolutionDetails<string>> {
    try {
      const result = await this.client.string(flagKey, this.toEvalContext(context));
      return { value: result, variant: result, reason: 'TARGETING_MATCH' };
    } catch (err) {
      return {
        value: defaultValue,
        reason: 'ERROR',
        errorCode: ErrorCode.GENERAL,
        errorMessage: err instanceof Error ? err.message : 'Unknown error',
      };
    }
  }

  async resolveNumberEvaluation(
    flagKey: string,
    defaultValue: number,
    context: EvaluationContext,
    _logger: Logger,
  ): Promise<ResolutionDetails<number>> {
    try {
      const result = await this.client.string(flagKey, this.toEvalContext(context));
      const num = Number(result);
      if (Number.isNaN(num)) {
        return { value: defaultValue, reason: 'ERROR', errorCode: ErrorCode.PARSE_ERROR, errorMessage: `Cannot parse "${result}" as number` };
      }
      return { value: num, variant: result, reason: 'TARGETING_MATCH' };
    } catch (err) {
      return {
        value: defaultValue,
        reason: 'ERROR',
        errorCode: ErrorCode.GENERAL,
        errorMessage: err instanceof Error ? err.message : 'Unknown error',
      };
    }
  }

  async resolveObjectEvaluation<T extends JsonValue>(
    flagKey: string,
    defaultValue: T,
    context: EvaluationContext,
    _logger: Logger,
  ): Promise<ResolutionDetails<T>> {
    try {
      const result = await this.client.string(flagKey, this.toEvalContext(context));
      const parsed = JSON.parse(result) as T;
      return { value: parsed, variant: result, reason: 'TARGETING_MATCH' };
    } catch (err) {
      return {
        value: defaultValue,
        reason: 'ERROR',
        errorCode: ErrorCode.GENERAL,
        errorMessage: err instanceof Error ? err.message : 'Unknown error',
      };
    }
  }
}
